File: crypto/src/test/cplusplus/ajtaicommitment.cpp
/*
 * Copyright (c) 2024-2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

#include <boost/test/unit_test.hpp>

#include "ajtaicommitment.h"
#include "poseidon2solinas62.h"
#include "solinas62.h"
#include "solinas62extension.h"

using namespace blacknet::crypto;

BOOST_AUTO_TEST_SUITE(AjtaiCommitments)

using Z = Solinas62Ring;
using R = Solinas62RingDegree64NTT;

BOOST_AUTO_TEST_CASE(Zq) {
    using CS = AjtaiCommitment<Z, NormP::Infinity>;
    auto drg = Poseidon2Solinas62Sponge<{1, 2, 3, 4}>();
    CS cs(CS::setup(drg, 2, 2), 8);
    Z z1(1);
    Z z2(2);
    Z z3(3);
    Z z4(4);
    auto c1 = cs.commit({z1, z2});
    auto c2 = cs.commit({z3, z4});
    BOOST_TEST(cs.open(c1, {z1, z2}), "Opening");
    BOOST_TEST(!cs.open(c2, {z1, z2}), "Binding");
    BOOST_TEST(!cs.open(c1, {z2, z1}), "Positional binding");
    BOOST_TEST(cs.open(c1 + c2, {z1 + z3, z2 + z4}), "Homomorphism");
}

BOOST_AUTO_TEST_CASE(Rq) {
    using CS = AjtaiCommitment<R, NormP::Infinity>;
    auto drg = Poseidon2Solinas62Sponge<{5, 6, 7, 8}>();
    CS cs(CS::setup(drg, 2, 2), 16);
    R r1({1, 2});
    R r2({3, 4});
    R r3({5, 6});
    R r4({7, 8});
    auto c1 = cs.commit({r1, r2});
    auto c2 = cs.commit({r3, r4});
    BOOST_TEST(cs.open(c1, {r1, r2}), "Opening");
    BOOST_TEST(!cs.open(c2, {r1, r2}), "Binding");
    BOOST_TEST(!cs.open(c1, {r2, r1}), "Positional binding");
    BOOST_TEST(cs.open(c1 + c2, {r1 + r3, r2 + r4}), "Homomorphism");
}

BOOST_AUTO_TEST_SUITE_END()

File: crypto/src/test/rust/ajtaicommitment.rs
/*
 * Copyright (c) 2024-2026 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use blacknet_crypto::ajtaicommitment::AjtaiCommitment;
use blacknet_crypto::algebra::IntegerRing;
use blacknet_crypto::commitmentscheme::CommitmentScheme;
use blacknet_crypto::matrix::{DenseMatrix, DenseVector};
use blacknet_crypto::norm::{L2, LInf, NormBound};

type Z = blacknet_crypto::pervushin::PervushinField;

#[test]
fn test() {
    let setup = DenseMatrix::new(
        2,
        2,
        [
            0x27266DA3020839D0u64 as i64,
            0x82C5945C2F27D053u64 as i64,
            0xCBC815FE12387BF6u64 as i64,
            0xC006EACA4C2362E9u64 as i64,
        ]
        .map(Z::new)
        .into(),
    );
    let b_ecd = NormBound::<L2, f64>::new(7.22);
    let b_inf = NormBound::<LInf, <Z as IntegerRing>::Int>::new(8);
    let cs_ecd = AjtaiCommitment::new(setup.clone(), b_ecd);
    let cs_inf = AjtaiCommitment::new(setup, b_inf);
    let z1 = Z::new(1);
    let z2 = Z::new(2);
    let z3 = Z::new(3);
    let z4 = Z::new(4);
    let m12 = DenseVector::from([z1, z2]);
    let m21 = DenseVector::from([z2, z1]);
    let m34 = DenseVector::from([z3, z4]);
    let (c12, _) = cs_inf.commit(&m12);
    let (c34, _) = cs_inf.commit(&m34);

    assert!(cs_inf.open(&c12, &m12, &()), "Opening");
    assert!(!cs_inf.open(&c34, &m12, &()), "Binding");
    assert!(!cs_inf.open(&c12, &m21, &()), "Positional binding");
    assert!(
        cs_inf.open(&(&c12 + &c34), &(&m12 + &m34), &()),
        "Homomorphism"
    );

    assert!(cs_ecd.open(&c12, &m12, &()), "Opening");
    assert!(!cs_ecd.open(&c34, &m12, &()), "Binding");
    assert!(!cs_ecd.open(&c12, &m21, &()), "Positional binding");
    assert!(
        cs_ecd.open(&(&c12 + &c34), &(&m12 + &m34), &()),
        "Homomorphism"
    );
}
